### PR TITLE
Fix missing EOS penalties for XPO and Nash-MD

### LINKED ISSUE #6805: XPO/Nash-MD: missing_eos_penalty checks the full sequence including the prompt, never applies with EOS-in-prompt chat templates
### Description

In the experimental XPO and Nash-MD trainers, `missing_eos_penalty` checks for the EOS token over the **full sequence, including the prompt**:

```python
# trl/experimental/xpo/xpo_trainer.py (same pattern in nash_md_trainer.py)
if self.args.missing_eos_penalty is not None:
    model_contain_eos = torch.any(model_data["input_ids"] == self.processing_class.eos_token_id, dim=-1)
    ref_contain_eos = torch.any(ref_data["input_ids"] == self.processing_class.eos_token_id, dim=-1)
    model_scores[~model_contain_eos] -= self.args.missing_eos_penalty
    ref_scores[~ref_contain_eos] -= self.args.missing_eos_penalty
```

`model_data["input_ids"]` is the concatenated prompt + completion. For chat templates whose prompt already contains the EOS token — e.g. Qwen, where `<|im_end|>` is both the end-of-turn marker in the prompt and the EOS token — `torch.any(... == eos)` is always `True`, so the penalty **silently never applies**, regardless of whether the completion actually terminated.

The reward-model scoring right above already has `context_length` in scope, and the sibling trainers do this correctly:

- Online DPO checks `completion_ids` after `truncate_right` (`val/contain_eos_token` and penalty are completion-based).
- PPO checks `postprocessed_response` only.

So this looks like an accidental divergence in the XPO/Nash-MD copies rather than a deliberate choice. The fix is to slice to the completion, e.g.:

```python
model_contain_eos = torch.any(
    model_data["input_ids"][:, context_length:] == self.processing_class.eos_token_id, dim=-1
)
```

Affected locations (at f72460224c5477e65031b2363f54368699105182):
- `trl/experimental/xpo/xpo_trainer.py`, `_compute_rewards` (~lines 264–270)
- `trl/experimental/nash_md/nash_md_trainer.py`, `_compute_rewards` (~lines 333–339)

Related: #3750 raised a similar concern for the PPO trainer, but the XPO/Nash-MD copies were never covered.

I'm happy to submit a PR.

### Reproduction

Code inspection; the behavior follows directly from the excerpt above. Any XPO/Nash-MD run with a reward model, `missing_eos_penalty` set, and a chat template that emits EOS in the prompt (e.g. Qwen2.5-Instruct) never applies the penalty: truncated completions score the same as terminated ones.

### System Info

Found by code inspection (AI-assisted audit) at f72460224c5477e65031b2363f54368699105182.

- Platform: macOS-26.5.1-arm64-arm-64bit
- Python version: 3.12.7
- TRL version: 1.5.1+f724602
- PyTorch version: 2.7.1
- Transformers version: 5.8.1
- Accelerate version: 1.8.1

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by guptaishaan**
# What does this PR do?

Fixes #6805

XPO and Nash-MD checked the full prompt plus completion for EOS before applying `missing_eos_penalty`. This change limits the check to completion tokens, so an EOS token in a chat-formatted prompt cannot hide a missing completion EOS.

Regression tests cover both trainers with EOS in the prompt and completions with and without EOS. Both experimental trainer test files passed on Linux using CPU execution, with 17 tests passed and 6 skipped. Ruff, formatting, and doc-builder style checks passed.

GPU execution and full-size Qwen training were not verified because the installed PyTorch build required a newer CUDA driver than the available driver. The full repository test suite was not run.

Thanks to @mmjerge for reporting the issue and identifying the affected trainers.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Changes reward shaping in experimental RLHF trainers when missing_eos_penalty is enabled; behavior shifts for chat-formatted prompts but matches intended semantics and is covered by new unit tests.
> 
> **Overview**
> Fixes **missing EOS penalty** behavior in **XPO** and **Nash-MD** when `missing_eos_penalty` is set. `_compute_rewards` used to treat an EOS anywhere in `input_ids` (prompt + completion) as satisfying the check, so chat-style prompts that already contain EOS could skip the penalty even when the generated completion had no EOS.
> 
> The EOS check is now limited to **completion tokens** via `input_ids[:, context_length:]`, aligned with how `_log_statistics` already measures EOS. Regression tests call `_compute_rewards` with mocked rewards and assert the penalty applies when only the prompt has EOS but the completion does not.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8b4fb60bb088f691788f5c0e6102c5f9b28398e3. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by guptaishaan** [comment]
No need to close it manually. The PR has `Fixes #6805`, so GitHub will close the issue automatically when the PR is merged. Thanks!
